java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java

```diff
@@ -0,0 +1,180 @@
+import org.sonar.check.Rule;
+import org.sonar.java.model.DefaultJavaFileScannerContext;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.ConstraintManager;
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+import org.sonar.java.se.ObjectConstraint;
+import org.sonar.java.se.ProgramState;
+import org.sonar.java.se.SymbolicValue;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2222",
+  name = "Locks should be released",
+  priority = Priority.CRITICAL,
+  tags = {"bug", "cwe", "multi-threading"})
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.SYNCHRONIZATION_RELIABILITY)
+@SqaleConstantRemediation("20min")
+public class LocksNotUnlockedCheck extends SECheck implements JavaFileScanner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LocksNotUnlockedCheck.class);
+
+  private static enum States {
+    LOCKED, UNLOCKED;
+  }
+
+  private static class TransitionFactory {
+    private final ExpressionTree expression;
+    private final SymbolicValue value;
+
+    TransitionFactory(ExpressionTree transitionExpression, SymbolicValue instanceValue) {
+      expression = transitionExpression;
+      value = instanceValue;
+    }
+
+    ProgramState lockedState(ProgramState seed, SymbolicValue result) {
+      final ProgramState state = seed.addConstraint(result, BooleanConstraint.TRUE);
+      return state.addConstraint(value, new ObjectConstraint(expression, States.LOCKED));
+    }
+
+    ProgramState unlockedState(ProgramState seed, SymbolicValue result) {
+      final ProgramState state = seed.addConstraint(result, BooleanConstraint.FALSE);
+      return state.addConstraint(value, new ObjectConstraint(expression, States.UNLOCKED));
+    }
+  }
+
+  private static class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private static final String LOCK = "java.util.concurrent.locks.Lock";
+    private static final String LOCK_METHOD_NAME = "lock";
+    private static final String TRY_LOCK_METHOD_NAME = "tryLock";
+    private static final String UNLOCK_METHOD_NAME = "unlock";
+
+    private TransitionFactory transitionFactory = null;
+
+    public PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    TransitionFactory transitionFactory() {
```

Comment:
As you are working in a private static inner class I am really unsure about the pertinence of a getter in this case.